### ISSUE TITLE
Big endian problems?

### ISSUE THREAD

**Issue by ibara** [FascinatedBox/lily#327]
Hi --

On OpenBSD/octeon (64-bit big endian MIPS CPU), building lily is successful.
However, running the pre-commit-tests program fails rather quickly with:
```
SyntaxError: Class ? has no method or property named format.
    from test/test.lily:45:
```
For comparison, on OpenBSD/amd64, all the tests pass.

Please let me know what I can do to help. I intend to add lily to the OpenBSD packages repository and would like to have it available for all our architectures, not just the little endian ones.

**Comment by FascinatedBox** [comment]
To begin with, I appreciate the interest in wanting to port the language o another platform.

The line in question is this: `var result = "{0}".format(e)`. This is a situation that absolutely shouldn't be happening, because `String.format` should be pulling type information by having the `String` literal being run first. I'm not even sure how this would be happening, so I'd need to poke around for a while to investigate. Supposing that Travis has that platform available, I can create a different repo and play around with it for a while.

Before d9c5f17, Lily's testing suite was run by pre-commit-hook.py. Can you do a checkout to that revision and run it with `--verbose`? The output is going to be pretty long, but since it won't be driven by Lily, I'll have an idea of how bad the problem is since each tests is done by being run from Python. That's quite a few revisions back, but I suspect the results will be the same since I don't think that area has been touched in quite some time.

**Comment by ibara** [comment]
If I'm reading Travis's documentation correctly, they only offer amd64 machines (some Ubuntu x86-64 and Mac OS X 10.10-10.12) so unfortunately it seems that Travis isn't helpful in finding endian bugs :(

I will run the test on my octeon machine and report back here with the log when it's finished.

**Comment by ibara** [comment]
Hi again --

I had to roll back to https://github.com/FascinatedBox/lily/commit/b9018adba0d2df040677b8eb371a45d3e3757d6c because that was the last to have the `pre-commit-hook.py` file.

I built lily with `-DDEBUG=1` given to `cmake` so that it would get the `-O0` and `-g3` flags in case a core dump was generated during testing.

Attached is the log generated. I ended up having to issue `^C` because `lily` just spun at one point. I have it a full half hour of CPU time to try to resolve itself but it never did.

I also do have a core dump if there is any information you would like me to get from it.

Log file: [lily-octeon.txt](https://github.com/FascinatedBox/lily/files/1171305/lily-octeon.txt)

**Comment by FascinatedBox** [comment]
Wow this is bad. I was hoping it was just getting confused by `String.format` having an unusual type. But no, it seems to be confusing `String` for `?` in a lot of places.

I suspect the bad free and closure transform are intertwined somehow. There are a few areas in Lily where a 32-bit number is truncated down to a 16-bit one (particularly with line numbers). I'm really, really hoping that the closure issue is some actually easy underlying issue.

If you've got some time, perhaps we could talk this over on the discord? I think that once these problems are solved that they won't come back, and the thought of Lily being available on another platform is quite appealing.

**Comment by FascinatedBox** [comment]
When time permits, can you take another swing at this and see if it's still a problem? I think I fixed the underlying problem, but there's been a lot of change in the language that may have caused a different issue to have cropped up.

I think I changed the testing suite around the time that this was filed. The testing suite can now be invoked as:

```
cmake .
make
./pre-commit-tests
```

**Comment by FascinatedBox** [comment]
Don't worry about it. In the meantime I managed to get an emulator for bad mips working.

I did some digging around to discover what was going on. The commit that said it fixed the problem was in the right area. But it's not the right solution.

It's the use of class IDs in the lower 16 and flags in the upper 16 of the same 32 thats causing it.

Fixing it is doable, but it requires adjusting areas like printing, gc, and so on that dispatch on class id.

I'm going to try to push a fix for this before the next release. I'll verify that the emulator on my side is working before I poke you again to verify that this is done once and for all.
